### ISSUE TITLE
Provide a way to mark module-level attributes as being deprecated

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3937]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)| @jonathanj reported|
|-|-|
|Trac ID|trac#3937|
|Type|enhancement|
|Created|2009-07-29 10:19:27Z|
|Branch|https://github.com/twisted/twisted/tree/module-attribute-deprecation-3937|

It would be useful to be able to declare module-level attributes as being deprecated. Currently `twisted.python.deprecate` provides no way of doing this.

<details><summary>Searchable metadata</summary>

```
trac-id__3937 3937
type__enhancement enhancement
reporter__jonathanj jonathanj
priority__normal normal
milestone__ 
branch__branches_module_attribute_deprecation_3937 branches/module-attribute-deprecation-3937
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1248862767000000 1248862767000000
changetime__1257840789000000 1257840789000000
version__None None
owner__jonathanj jonathanj

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)|@jonathanj removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve set owner to @jonathanj|
|-|-|

Thanks for the branch. Some comments: 
 * tests are not passing under Python2.3
 * `_ModuleProxy` should have a public method for managing `_lazyDefinitions`, so that `deprecatedModuleAttributes` doesn't have to access it directly.

Everything else looks good. I'm a bit frighten by that namedAny call, but I can't really find a problem.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

In addition to therve's comments:

  1. I suggest a better API for the `getDeprecationWarningString` API change: `format` instead of `message`.  Interpolate the information into the string given; default to a string which produces the current results.  This is more flexible and less code.
  1. `ModuleProxyTests` is missing a docstring
  1. `DeprecatedAttributeTests` is missing a docstring
  1. `inspect.currentframe` is a much nicer way to spell `sys._getframe(0)`
  1. The various `addStackLevel` functions are obscure, uncommented, and in test methods which don't discuss the call stack depth at all.  What the heck?
  1. `_DeprecatedAttribute` is missing a docstring.
  1. What's wrong with `isinstance`?  `deprecatedModuleAttributes` grossly does an identity comparison on two type objects.
  1. I would spend some time contemplating whether it is really worthwhile to be able to present an API which doesn't require a module name if doing so requires messing with `sys._getframe`.  If the tests changed to simply import the test module, then they wouldn't need to use this private implementation-specific function, either.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_4"></a>|@jonathanj commented|
|-|-|

(In [27236]) Improve the new API for getDeprecationWarningString, remove magic stack
tomfoolery and improve / add documentation.

refs [#3937](https://github.com/twisted/twisted/issues/3937)

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_5"></a>|@jonathanj set owner to @exarkun|
|-|-|

Replying to [therve](#note_2):
>  * tests are not passing under Python2.3

This should hopefully be fixed after removing all the frame / stack magic.

>  * `_ModuleProxy` should have a public method for managing `_lazyDefinitions`, so that `deprecatedModuleAttributes` doesn't have to access it directly.

Done.

----

Replying to [exarkun](#note_3):
>   1. I suggest a better API for the `getDeprecationWarningString` API change: `format` instead of `message`.  Interpolate the information into the string given; default to a string which produces the current results.  This is more flexible and less code.

Good idea, done.

>   4. `inspect.currentframe` is a much nicer way to spell `sys._getframe(0)`

>   5. The various `addStackLevel` functions are obscure, uncommented, and in test methods which don't discuss the call stack depth at all.  What the heck?

>   8. I would spend some time contemplating whether it is really worthwhile to be able to present an API which doesn't require a module name if doing so requires messing with `sys._getframe`.  If the tests changed to simply import the test module, then they wouldn't need to use this private implementation-specific function, either.

Most (except for one, that is now commented) of the `addStackLevel` stuff is gone and all of the frame magic has been replaced by a module name parameter.

I addressed the rest of your comments too.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_6"></a>|@jonathanj commented|
|-|-|

(In [27237]) Add missing module.

refs [#3937](https://github.com/twisted/twisted/issues/3937)

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_7"></a>|@jonathanj commented|
|-|-|

(In [27238]) Document the fact that "__name__" is a useful way to address the current
module, when deprecating attributes.

refs [#3937](https://github.com/twisted/twisted/issues/3937)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @jonathanj|
|-|-|

  1. I feel like `deprecationWarningFormat` is a "constant" and should be `DEPRECATION_WARNING_FORMAT`.
  1. `_getDeprecationWarningString` documents a possible type of its `format` parameter as `None`, but that's not a type.
  1. `_ModuleProxy` won't proxy a number of attributes; `_module`, `_lazyAttributes`, `deprecateAttribute`, etc.  The private attributes probably aren't _hugely_ concerning.  Most code shouldn't be accessing private names of a module anyway.... But unit tests for that module are probably allowed to.  And anyway, `deprecateAttribute` isn't private, so it should really be proxied correctly.  Using `__getattribute__` instead will let the proxy proxy everything.  This breaks the strategy of caching attributes on the proxy, though.  So how important is the performance of getting attributes from a module that has deprecated attributes?  And what is the performance difference between `__getattribute__` and the current `__getattr__`-based implementation?  Do we care?
  1. Any particular reason you picked `namedAny` to implement `_DeprecatedAttribute.get`?  It's probably fine, but I probably would have gone with a getattr() on the wrapped module.  These have different semantics, but hopefully only when reload() is being used, so it doesn't really matter.
  1. `deprecatedModuleAttributes` doesn't document its `names` parameter.
  1. I don't think there's any test coverage for passing multiple names to that function.  Since the message should probably be different for most attributes anyway, I'm not sure there's much reason to support deprecating multiple names at once, anyway.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_9"></a>|@jonathanj commented|
|-|-|

(In [27240]) Use a "__getattribute__" approach for the module proxy and stop supporting
deprecation of multiple names at once.

Other more minor changes include renaming some constants, documentation and
tests.

refs [#3937](https://github.com/twisted/twisted/issues/3937)

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_10"></a>|@jonathanj set owner to @exarkun|
|-|-|

Replying to [exarkun](#note_8):
>   3. `_ModuleProxy` won't proxy a number of attributes; `_module`, `_lazyAttributes`, `deprecateAttribute`, etc.  The private attributes probably aren't _hugely_ concerning.  Most code shouldn't be accessing private names of a module anyway.... But unit tests for that module are probably allowed to.  And anyway, `deprecateAttribute` isn't private, so it should really be proxied correctly.  Using `__getattribute__` instead will let the proxy proxy everything.  This breaks the strategy of caching attributes on the proxy, though.  So how important is the performance of getting attributes from a module that has deprecated attributes?  And what is the performance difference between `__getattribute__` and the current `__getattr__`-based implementation?  Do we care?

This seems like a good proposal, I don't think the performance impact warrants enough caring to not do this:

```
$ python -m timeit -s 'import isdeprecated as go' 'go.DEPRECATED_ATTRIBUTE'
/usr/lib/python2.5/timeit.py:6: DeprecationWarning: isdeprecated.DEPRECATED_ATTRIBUTE was deprecated in Twisted 8.0.0: Oh noes!
100000 loops, best of 3: 10.4 usec per loop

$ python -m timeit -s 'import isdeprecated as go' 'go.ATTRIBUTE'
100000 loops, best of 3: 9.99 usec per loop

$ python -m timeit -s 'import notdeprecated as go' 'go.ATTRIBUTE'
1000000 loops, best of 3: 0.304 usec per loop
```

>   6. I don't think there's any test coverage for passing multiple names to that function.  Since the message should probably be different for most attributes anyway, I'm not sure there's much reason to support deprecating multiple names at once, anyway.

I still like the idea of being able to deprecate multiple names at once. However, deprecating synonyms are the only use-case I can come up with, which probably means it isn't worth supporting.

All other commentary was addressed too.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @jonathanj|
|-|-|

  1. `ModuleProxyTests.test_getattrPassthrough` has a typo for `AttributeError` in it.
  1. Please avoid `assertEqual` and use `assertEquals` everywhere.
  1. `ModuleProxyTests.test_setattr` could directly assert that the attribute is set on the wrapped module.  Also, isn't this test leaking an attribute onto `twisted.python.deprecate`?  Using a throw-away object instead might be better.
  1. `DeprecatedAttributeTests.test_deprecatedAttributeHelper` should probably assert something about the warning which was emitted.  Its category, at least, and perhaps its message as well.
  1. Similarly for `DeprecatedAttributeTests.test_deprecatedAttribute`
  1. Period missing at the end of the `@var` description of `DEPRECATION_WARNING_FORMAT` in `deprecate.py`.
  1. There is a commented out `setattr` call in `_ModuleProxy.__getattribute__`.  It looks like this is actually necessary for an attribute to be looked up more than once, but `test_deprecatedAttribute` isn't catching the problem, perhaps because of the weird way it uses `deprecatedModuleAttribute` (ie, trying to deprecate an attribute that's already deprecated, and doing so after the module has been imported).
  1. `_ModuleProxy`'s methods, `__repr__`, `__setattr__`, and `__getattribute__`, need docstrings.
  1. `_ModuleProxy.__repr__` is untested

Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_12"></a>|@jonathanj commented|
|-|-|

(In [27399]) Assert some things about the emitted deprecation warnings. refs [#3937](https://github.com/twisted/twisted/issues/3937)

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_13"></a>|@jonathanj commented|
|-|-|

(In [27404]) Issue warnings for every access to deprecated attributes. Improve documentation and test coverage. refs [#3937](https://github.com/twisted/twisted/issues/3937)

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_14"></a>|@jonathanj commented|
|-|-|

(In [27407]) Use a throw-away object for ModuleProxyTests. refs [#3937](https://github.com/twisted/twisted/issues/3937)

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_15"></a>|@jonathanj set owner to @exarkun|
|-|-|

Sorry about the long delay between the last review and any additional changes.

All of the review points have been addressed.

Regarding point 7: This issue actually went away by itself once I changed deprecated attributes to always issue a warning (the warning module can filter them) when accessed, instead of only the first time.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun set owner to @jonathanj<br>@exarkun set status to `new`|
|-|-|

Great, thanks.  A few things:

  1. The `twisted.python.deprecate` module docstring still talks about warnings being issued only the first time.
  1. There's some trailing whitespace in one of the tests, `test_getattrIntercept`
  1. I think there's a test missing for the `isinstance` check in `deprecatedModuleAttribute`
  1. The use of `getattr` in `test_deprecatedAttribute` is a little weird.  Is it just that way so the deprecated attribute name doesn't have to be repeated?  If so, I'd say that in a comment.  My first thought was that `getattr` was doing something to stack levels to make the deprecation warning come out the right place.

Basically looks good.  If you don't run into any surprises addressing these points, please go ahead and merge.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_18"></a>|@jonathanj set owner to @exarkun|
|-|-|

I'm putting this back into review just for some additional input on my solution for your third point.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun set owner to @jonathanj|
|-|-|

looks good, except you really want to do the `addCleanup` in `test_wrappedModule` immediately after you add the module to `sys.modules`, before you do anything that might fail.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_20"></a>|@jonathanj set status to `closed`|
|-|-|

(In [27422]) Merge module-attribute-deprecation-3937.

Author: jonathanj
Reviewer: exarkun
Fixes: [#3937](https://github.com/twisted/twisted/issues/3937)

Add a function to twisted.python.deprecate that enables the deprecation of
module-level attributes which, when accessed, issue a deprecation warning.
